tests/integration/test_charm_upgrade.py

```diff
@@ -3,35 +3,65 @@
 
 """Integration tests for charm upgrades."""
 
+import functools
+import pathlib
+
 import pytest
 from juju.client import client
 from juju.model import Model
+from pytest_operator.plugin import OpsTest
 
-from charm_state import TEST_MODE_CONFIG_NAME, VIRTUAL_MACHINES_CONFIG_NAME
-from tests.integration.helpers import deploy_github_runner_charm
+from charm_state import VIRTUAL_MACHINES_CONFIG_NAME
+from tests.integration.helpers import (
+    deploy_github_runner_charm,
+    inject_lxd_profile,
+    is_upgrade_charm_event_emitted,
+    wait_for,
+)
 
 
 @pytest.mark.asyncio
 async def test_charm_upgrade(
     model: Model,
+    ops_test: OpsTest,
     charm_file: str,
+    loop_device: str | None,
     app_name: str,
     path: str,
     token: str,
     http_proxy: str,
     https_proxy: str,
     no_proxy: str,
+    tmp_path: pathlib.Path,
 ):
     """
-    arrange: given latest stable version of the charm.
+    arrange: given latest stable version of the charm (current 161).
     act: charm upgrade is called.
     assert: the charm is upgraded successfully.
     """
-    latest_stable_revision = 161
+    latest_stable_path = tmp_path / "github-runner.charm"
+    latest_stable_revision = 161  # update this value every release to stable.
```

Comment:
I don't think this will be followed (update on very release :D) unless there is an implemented check letting the test fail.

Reply:
Lets see if this is required in further tests - currently it's just downloading from latest/stable jammy. This field was required to mark the original source of the charm during upgrade :)